lib/vsc/utils/ssh_util.py

```diff
@@ -0,0 +1,176 @@
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# the Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/hpcugent/vsc-utils
+#
+# vsc-utils is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Library General Public License as
+# published by the Free Software Foundation, either version 2 of
+# the License, or (at your option) any later version.
+#
+# vsc-utils is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Library General Public License for more details.
+#
+# You should have received a copy of the GNU Library General Public License
+# along with vsc-utils. If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Ssh helper class
+"""
+
+import os
+import logging
+import paramiko
+import socket
+
+class Ssh_util(object):
+    """Helper class for ssh connections."""
+
+    def __init__(self, host=None, user=None, passwd=None, port=22, agent_forwarding=False,
+                 ssh_config=False, connection_timeout=10, command_timeout=60):
+        """constructor"""
+        self.host = host
+        self.port = port
+
+        self.user=user
+        self.passwd = passwd
+
+        self.conn_timeout = connection_timeout
+        self.comm_timeout = command_timeout
+
+        self.ssh_config = ssh_config
+        self.agent_forw = agent_forwarding
+        self.proxy = None
+
+        self.client = None
+
+    def _load_user_config(self):
+        """Read proxy commands from local ssh config file."""
+        ssh_config_file = os.path.expanduser("~/.ssh/config")
+
+        if os.path.exists(ssh_config_file):
+            conf = paramiko.SSHConfig()
+            with open(ssh_config_file) as f:
+                conf.parse(f)
+
+            host_config = conf.lookup(self.host)
+            if 'proxycommand' in host_config:
+                self.proxy = paramiko.ProxyCommand(host_config['proxycommand'])
+
+    def connect(self):
+        """Create a connection to a host."""
+        self.client = paramiko.SSHClient()
+
+        logging.debug("Establishing ssh connection to %s", self.host)
+        if self.ssh_config:
+            self._load_user_config()
+
+        self.client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        self.client.load_system_host_keys()
+
+        try:
+            if self.passwd:
+                self.client.connect(hostname=self.host, username=self.user, timeout=self.conn_timeout,
+                                    sock=self.proxy, allow_agent=False)
+            else:
+                self.client.connect(hostname=self.host, timeout=self.conn_timeout, sock=self.proxy)
+
```

Comment:
`return True` here? Not sure.

Reply:
I know, but returning early is easier to see in this case. You do not do anything after the except clauses, so might as well return here.

Reply:
And then the return at the end can be False, avoiding all the separate returns in each except.